### PR TITLE
fix(storage): resolve GroupCommit data loss bugs and TieredStorage edge cases

### PR DESCRIPTION
**Pull request by madmax983**
This PR addresses critical correctness issues identified by Sentry in the storage layer.

### Group Commit Coordinator
- **Bug Fixed:** Previously, `GroupCommitCoordinator` used a single `last_flush_error` field. This caused two severe bugs:
  1. **False Success (Data Loss):** If Epoch 1 failed, but Epoch 2 succeeded immediately after, a waiter for Epoch 1 would wake up, see `flushed_epoch >= 1`, and find no error (because Epoch 2 cleared it). The transaction would return `Ok` despite data loss.
  2. **False Failure (Ghost Error):** If Epoch 1 succeeded, but Epoch 2 failed, a waiter for Epoch 1 might wake up, see `flushed_epoch >= 1`, and find the error from Epoch 2. The transaction would return `Err` despite success.
- **Fix:** Replaced `last_flush_error` with `recent_errors: VecDeque<(u64, String)>`. Waiters now check if *their specific epoch* is in the failure list. The list is bounded (1024 entries) to prevent memory leaks.
- **Verification:** Enabled ignored regression tests in `tests/repro_havoc_group_commit.rs`, which now pass.

### Tiered Storage Latency Tracker
- **Bug Fixed:** `LatencyTracker::record` would store 1 sample even if `max_samples` was 0, due to `pop_front` on an empty deque followed by `push_back`.
- **Fix:** Added an explicit check `if self.max_samples == 0 { return; }`.
- **Verification:** Added unit tests covering zero samples and percentile calculation for even sample sizes.

### Test Performance
- **Optimization:** `test_wait_for_flush_timeout` relied on the default 10s timeout. It now uses a custom `GroupCommitConfig` with a 10ms timeout, reducing test duration significantly.

---
*PR created automatically by Jules for task [9760664587579715059](https://jules.google.com/task/9760664587579715059) started by @madmax983*

### CODE REVIEW

File: src/storage/wal/group_commit.rs [hunk 1]
```diff
@@ -269,8 +273,15 @@ impl GroupCommitCoordinator {
             })
         })?;
 
-        // Store any error for propagation
-        state.last_flush_error = result.err().map(|e| e.to_string());
+        // Store error if any
+        if let Err(e) = result {
+            let epoch = state.current_epoch;
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/storage/wal/group_commit.rs]: > security-critical critical
> 
> The logic in `mark_flushed` uses `state.current_epoch` to record errors and then advances `flushed_epoch` to that same value. This creates a critical race condition: if a new transaction registers via `register_transaction` after a flush has started (and drained the WAL stripes) but before `mark_flushed` is called, it will be assigned the `current_epoch`. When `mark_flushed` completes, it will mark this epoch as flushed, even though the new transaction's data was not included in the batch. This leads to immediate data loss upon success reporting.
> 
> The `current_epoch` must be incremented _before_ the flush operation begins (e.g., in a `start_flush` method), ensuring that any transactions registering during the flush are assigned to the _next_ epoch. `mark_flushed` should then only update `flushed_epoch` and notify waiters for the completed epoch.

@jules

File: src/storage/wal/group_commit.rs [hunk 2]
```diff
@@ -236,11 +238,13 @@ impl GroupCommitCoordinator {
...
-                reason: format!("Group commit flush failed: {}", error_msg),
-            }));
+        // Check for flush errors specifically for this epoch
+        for (failed_epoch, error_msg) in &state.recent_errors {
+            if *failed_epoch == epoch {
+                return Err(Error::Storage(StorageError::WalError {
+                    reason: format!("Group commit flush failed: {}", error_msg),
+                }));
+            }
         }
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/storage/wal/group_commit.rs:256]: > security-high high
> 
> This section contains a "False Success" vulnerability. The `GroupCommitCoordinator` uses a bounded `VecDeque` (`recent_errors`) to store flush errors, but the check in `wait_for_flush` does not account for the possibility that an error might have been evicted from the history. If a transaction's flush fails, but subsequent flushes also fail before the transaction thread wakes up and checks the error list, the error for its epoch will have been popped from the deque. This results in a "False Success" where the transaction is reported as durable to the caller despite its data never having been successfully written to the WAL.
> 
> For example:
> 
> 1. The `recent_errors` capacity is 1024.
> 2. Epoch 100 fails. Its error is recorded.
> 3. A waiter for epoch 100 is delayed.
> 4. Epochs 101 through 1125 also fail, and their flushes complete. This causes the error for epoch 100 to be pushed out of the `VecDeque`.
> 5. The waiter for epoch 100 finally runs. It sees `flushed_epoch` is >= 100 and proceeds.
> 6. It checks `recent_errors`, doesn't find an entry for epoch 100, and returns `Ok(())`, even though its data was lost.
> 
> To ensure correctness and prevent this data loss, the coordinator should either maintain a larger or unbounded error history (with a different cleanup mechanism), or track the highest epoch that has been evicted from the error history. If a waiter's epoch is less than or equal to the highest evicted epoch and is not present in the current error list, it should be treated as a failure rather than a success. Specifically, we should check if the epoch we are waiting for is older than the oldest epoch present in the error history. If it is, we cannot know its status for sure, and we must return an error to prevent data loss.

@jules

File: src/storage/wal/group_commit.rs [hunk 3]
```diff
@@ -269,8 +273,15 @@ impl GroupCommitCoordinator {
...
-        // Store any error for propagation
-        state.last_flush_error = result.err().map(|e| e.to_string());
+        // Store error if any
+        if let Err(e) = result {
+            let epoch = state.current_epoch;
+            state.recent_errors.push_back((epoch, e.to_string()));
+            // Keep history limited
+            if state.recent_errors.len() > 1024 {
+                state.recent_errors.pop_front();
+            }
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/storage/wal/group_commit.rs]: > medium
> 
> The capacity of the `recent_errors` deque is hardcoded to `1024`. This magic number makes it difficult to configure and test edge cases, such as the one where an error record is evicted from the history.
> 
> To improve configurability and testability, I recommend making this capacity a parameter in `GroupCommitConfig`. This would also make the trade-off between memory usage and error history retention explicit to users of the coordinator. You would also need to add `pub recent_errors_capacity: usize` to `GroupCommitConfig` and initialize it accordingly.

@jules

File: src/storage/wal/group_commit.rs [hunk 4]
```diff
@@ -269,8 +273,15 @@ impl GroupCommitCoordinator {
...
 
-        // Store any error for propagation
-        state.last_flush_error = result.err().map(|e| e.to_string());
+        // Store error if any
+        if let Err(e) = result {
+            let epoch = state.current_epoch;
+            state.recent_errors.push_back((epoch, e.to_string()));
+            // Keep history limited
+            if state.recent_errors.len() > 1024 {
+                state.recent_errors.pop_front();
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/storage/wal/group_commit.rs]: > **P1 Badge Avoid evicting unconsumed epoch errors**
> 
> Bounding `recent_errors` to 1024 entries can discard a failed epoch before its transaction calls `wait_for_flush` (for example, if that thread is descheduled while the flush thread keeps advancing through repeated failures such as a disk-full loop). Once the `(epoch, error)` pair is evicted, `wait_for_flush` will observe `flushed_epoch >= epoch` and no matching error, incorrectly returning success for a failed flush, which reintroduces the false-success durability bug this change is intended to eliminate.
> 
> Useful? React with 👍 / 👎.

@jules
